File: archrules-common/src/main/java/com/netflix/nebula/archrules/common/KotlinInternalMethodPredicate.java
package com.netflix.nebula.archrules.common;

import com.tngtech.archunit.base.DescribedPredicate;
import com.tngtech.archunit.core.domain.JavaMethod;
import kotlin.Metadata;
import kotlin.metadata.Attributes;
import kotlin.metadata.KmFunction;
import kotlin.metadata.Visibility;
import kotlin.metadata.jvm.KotlinClassMetadata;

import java.util.List;
import java.util.Optional;

import static com.netflix.nebula.archrules.common.KotlinMetadataUtil.matchFunction;

class KotlinInternalMethodPredicate extends DescribedPredicate<JavaMethod> {

    KotlinInternalMethodPredicate() {
        super("Kotlin internal method");
    }

    @Override
    public boolean test(JavaMethod javaMethod) {
        if (javaMethod != null && javaMethod.getOwner().isAnnotatedWith("kotlin.Metadata")) {
            KotlinClassMetadata metadata = KotlinClassMetadata.readStrict(javaMethod.getOwner().getAnnotationOfType(Metadata.class));
            if (metadata instanceof KotlinClassMetadata.Class) {
                List<KmFunction> functions = ((KotlinClassMetadata.Class) metadata).getKmClass().getFunctions();
                Optional<KmFunction> matchingFunction = matchFunction(javaMethod, functions);
                return matchingFunction
                        .filter(it -> Attributes.getVisibility(it) == Visibility.INTERNAL)
                        .isPresent();
            } else if (metadata instanceof KotlinClassMetadata.FileFacade) {
                List<KmFunction> functions = ((KotlinClassMetadata.FileFacade) metadata).getKmPackage().getFunctions();
                Optional<KmFunction> matchingFunction = matchFunction(javaMethod, functions);
                return matchingFunction
                        .filter(it -> Attributes.getVisibility(it) == Visibility.INTERNAL)
                        .isPresent();
            } else {
                return false;
            }
        } else {
            return false;
        }
    }
}

File: archrules-common/src/main/kotlin/com/netflix/nebula/archrules/common/KotlinInternalMethodPredicate.kt
package com.netflix.nebula.archrules.common

import com.tngtech.archunit.base.DescribedPredicate
import kotlin.collections.mapIndexed
import kotlin.reflect.KVisibility
import kotlin.reflect.full.functions
import kotlin.reflect.full.valueParameters
import kotlin.reflect.jvm.javaMethod
import kotlin.reflect.jvm.jvmErasure

class KotlinInternalMethodPredicate : DescribedPredicate<com.tngtech.archunit.core.domain.JavaMethod>(
    "Kotlin internal method"
) {

    /**
     * @param javaMethod declared as nullable in order to avoid static references to the kotlin Intrinsics class
     */
    override fun test(javaMethod: com.tngtech.archunit.core.domain.JavaMethod?): Boolean {
        return javaMethod != null && if (javaMethod.owner.isAnnotatedWith("kotlin.Metadata")) {
            val functions = javaMethod.owner.reflect().kotlin.functions
            val matchingFunction = functions
                .firstOrNull {
                    it.javaMethod?.name == javaMethod.name && it.valueParameters.mapIndexed { index, parameter ->
                        parameter.type.jvmErasure.qualifiedName == javaMethod.parameters[index].type.toErasure().fullName
                    }.all { true }
                }
            matchingFunction != null && matchingFunction.visibility == KVisibility.INTERNAL
        } else {
            false
        }
    }
}
